core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java

```diff
@@ -280,37 +281,35 @@ protected void onTaskCompletion(InboundTransferTask inboundTransfer) {
 
          // With passivation, some key could be activated here and we could miss it,
          // but then it should be broadcast-loaded in PrefetchInvalidationInterceptor
-         AdvancedCacheLoader<Object, Object> stProvider = persistenceManager.getStateTransferProvider();
-         if (stProvider != null) {
-            try {
-               Flowable.fromPublisher(stProvider.entryPublisher(k -> !dataContainer.containsKey(k), true, true))
-                     .blockingForEach(me -> {
-                        int segmentId = keyPartitioner.getSegment(me.getKey());
-                        if (finalCompletedSegments.contains(segmentId)) {
-                           try {
-                              Metadata metadata = me.getMetadata();
-                              if (metadata instanceof RemoteMetadata) {
-                                 Address backup = ((RemoteMetadata) metadata).getAddress();
-                                 retrieveEntry(me.getKey(), backup);
-                                 for (Address member : cacheTopology.getActualMembers()) {
-                                    if (!member.equals(backup)) {
-                                       invalidate(me.getKey(), metadata.version(), member);
-                                    }
-                                 }
-                              } else {
-                                 backupEntry(entryFactory.create(me.getKey(), me.getValue(), me.getMetadata()));
-                                 for (Address member : nonBackupAddresses) {
-                                    invalidate(me.getKey(), metadata.version(), member);
-                                 }
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
-                           }
-                        }
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         Publisher<MarshallableEntry<Object, Object>> persistencePublisher =
+            persistenceManager.publishEntries(completedSegments, k -> !dataContainer.containsKey(k), true, true,
+                                              StoreConfiguration::fetchPersistentState);
+         try {
+            Flowable.fromPublisher(persistencePublisher)
+                    .blockingForEach(me -> {
```

Comment:
Actually now that I think about this, this is actually blocking. Not because of the invalidation - but because the persistence manager underneath can take time between publishing entries. We need to have a non blocking call here and wait for its completion.

A `blockingForEach` is essentially just a blocking iterator that invokes `forEachRemaining`. So if the entries are not returned in the same thread, we will end up blocking.

Reply:
I was writing about the `invalidate()` call, because I remember reading this code and thinking we really shouldn't send the entries/invalidations in a `blockingForEach()` call, but I see now I made it more confusing :)

I only meant to remove the usage of `persistenceManager.getStateTransferProvider()`, so I'll remove the comment. I'm going to have another look at the scattered state consumer in another PR, because I also want to limit the number of in-flight invalidation/backup commands.